enhancements/node-tuning/secure-shortname-image-pulls.md

```diff
@@ -0,0 +1,216 @@
+---
+title: secure-shortname-image-pulls
+authors:
+  - "@umohnani8"
+reviewers:
+  - "@mrunalp"
+approvers:
+  - "@mrunalp"
+api-approvers: # in case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers)
+  - N/A
+creation-date: 2022-01-26
+last-updated: 2022-01-26
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/RUN-1134
+see-also:
+replaces:
+superseded-by:
+---
+
+# Secure Shortname Image Pulls
+
+## Summary
+
+Pulling images via shortnames is not very secure. There is a possibility of
+spoofing where the image is pulled from as it goes through the list of `unqualified-search-registries`
+checking for a match and pulls the image from the first location it is found in. With this method
+there is no guarantee that the image is actually being pulled from the correct source.
+
+We have made this more secure by adding support for an alias table that has a list of shortname images
+pointing to their fully qualified image name. This alias table is checked for a match whenever a pull
+via shortname is attempted, which guarantees that the image will only be pulled from the location specified
+in the alias table.
+
+## Motivation
+
+We are always trying to make OpenShift more secure. While we do highly discourage the usage of
+shortnames in favour of fully qualified names, there are users who still prefer using shortnames
+so we have updated the pull logic to be more secure than just checking a list of registries and pulling
+from the first location where a match is found.
+
+So, to make the shortname use-case more secure, the c/image library was updated to check an
+alias table when available. An alias table is a list of image names with shortnames pointing to their
+corresponding fully qualified image names. We can use this new feature to ensure that any Red Hat and OpenShift
+images that are referred to by short name will always resolve to our official locations, hence
+guaranteeing that those images will be official.
+
+### Goals
+
+- Secure where the official Red Hat and OpenShift images are pulled from when using shortnames
+- The alias table list will be maintained by us and be specific to the OpenShift use-case
```

Comment:
Will this be 100% automated, or does that involve an on-going manual process to update the lists, that will need to be regularly performed? Is there an established mechanism for doing that?

---

How will the list be shipped? And if it is inside a container image, like the MCO one, are there any concerns WRT relative timing of changes to the “ground truth” (e.g. adding a new official repo) vs. timing of creating/publishing/deploying updates to that list (e.g. on a long-term-support branch of MCO that only rarely receives updates, and those are deployed even more rarely by a customer)?

(The existing CoreOS-shipped list that is built from a RPM has very similar timing issues, of course — it’s a matter of what “specific to OpenShift” means and whether it’s something we can broadly guarantee to be useful in some specific ways, or whether it’s best thought of as a crutch for long-term widely-known images only.)

Reply:
The MCO can lay down files on the nod during bootstrap, so the plan here is to lay down an `00-alias-table.conf` under `/var/lib/containers/registries.conf.d` similar to how we lay down the the crio config file under crio.conf.d that overrides some of the crio configurations from the defaults we ship upstream with cri-o - https://github.com/openshift/machine-config-operator/blob/master/templates/master/01-master-container-runtime/_base/files/crio.yaml. When OpenShift is updated and the MCO updates, it will update any changes made to the files under templates as well - I am not seeing a timing issue here for that as it has worked for the other config files we lay down via templates, but please let me know if I am missing something.

As for how we will update the alias table, I think it might have to be a manual process initially till we figure out if there is a good way to automate it and where to get the aliases and their matching fully qualified names from for the OCP specific use case.

Reply:
The timing issue I worry about is:

- redhat.io/$product/$repo1 , and redhat.io/$product/$repo2 previously existed.
- We add them as aliases $product/$repo1 and $product/$repo2
- Customers start using those aliases, and rely on them existing.

(time passes)

- The product in question introduces redhat.io/$product/repo3 , and announces it.
- Customers, habitually, refer to $product/$repo3 , expecting an alias. This does not work because the alias doesn’t exist. (Customers, presumably, grumpily use a fully-qualified name, or define their own alias, or in the worst case, change USR.)

(time passes)

- A MCO update ships a $product/$repo3 alias. But, by then, no-one needs it to exist and it might even, strictly speaking, break something.

Reply:
> where to get the aliases and their matching fully qualified names from for the OCP specific use case.

BTW is there already some conceptual idea of what the ”OCP specific use case” list would be? I.e. how does it differ from the RHEL list, if it does differ at all? (Note that the alias in https://bugzilla.redhat.com/show_bug.cgi?id=2021026 is an `openshift3/…` alias; would that be included or excluded?)

Without articulating how that list differs it’s hard to me to think whether we need it.